### PR TITLE
added numpy.r_ and its test

### LINKED ISSUE #20814: c_

### LINKED ISSUE #20815: r_

### PR DESCRIPTION
**Pull request by ShreyanshBardia**
closes #20815
closes #20814

### CODE REVIEW

**Comment by ShreyanshBardia** [comment]
@sherry30 can you review this please

File: ivy/functional/frontends/numpy/__init__.py [hunk 1]
```diff
@@ -490,6 +489,8 @@ def promote_types_of_numpy_inputs(
 from .matrix import *
 from . import random
 from .random import *
+from . import indexing_routines
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/numpy/__init__.py:492]: This import should be left where it was as this is an unnecessary change right?🙂

File: ivy/functional/frontends/numpy/indexing_routines/inserting_data_into_arrays.py [hunk 2]
```diff
@@ -28,3 +30,138 @@ def fill_diagonal(a, val, wrap=False):
...
+                if step is None:
+                    step = 1
+                if ivy.is_complex_dtype(step):
+                    size = int(abs(step))
+                    newobj = np_frontend.linspace(start, stop, num=size).ivy_array
+                else:
+                    newobj = np_frontend.arange(start, stop, step).ivy_array
+                if ndmin > 1:
+                    newobj = np_frontend.array(
+                        newobj, copy=False, ndmin=ndmin
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/numpy/indexing_routines/inserting_data_into_arrays.py:77]: Here can we try to call `self.ndim` instead of creating a copy of it🙂

File: ivy/functional/frontends/numpy/indexing_routines/inserting_data_into_arrays.py [hunk 3]
```diff
@@ -28,3 +30,138 @@ def fill_diagonal(a, val, wrap=False):
...
+                res = res.T
+        return res
+
+    def __len__(self):
+        return 0
+
+
+class RClass(AxisConcatenator):
+    def __init__(self):
+        AxisConcatenator.__init__(self, 0)
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/numpy/indexing_routines/inserting_data_into_arrays.py:156]: We can use `super().__init__()` to conform with best inheritance pratices

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey @ShreyanshBardia 

Great work so far🫡
Below are a few comments to start working on and also would you update your branch with master branch for CI checks🙂

Thanks

File: ivy/functional/frontends/numpy/__init__.py (hunk 1 above)
Comment (ShreyanshBardia) [reply ivy/functional/frontends/numpy/__init__.py:492]: Hi @hirwa-nshuti, this was to avoid circular imports

File: ivy/functional/frontends/numpy/indexing_routines/inserting_data_into_arrays.py (hunk 3 above)
Comment (ShreyanshBardia) [reply ivy/functional/frontends/numpy/indexing_routines/inserting_data_into_arrays.py]: yes right, done!

File: ivy/functional/frontends/numpy/indexing_routines/inserting_data_into_arrays.py (hunk 2 above)
Comment (ShreyanshBardia) [reply ivy/functional/frontends/numpy/indexing_routines/inserting_data_into_arrays.py:77]: Not sure what you mean by that. ndmin is basically a parameter for minimum number of dimensions an array should be. If I am missing something can you please elaborate a bit @hirwa-nshuti

Comment (hirwa-nshuti) [reply ivy/functional/frontends/numpy/indexing_routines/inserting_data_into_arrays.py:77]: What I meant is that this should we can still reference to `self.ndim` as long as the value remains unchanged instead of creating new copies of such same variable🙄

Comment (ShreyanshBardia) [reply ivy/functional/frontends/numpy/indexing_routines/inserting_data_into_arrays.py:77]: @hirwa-nshuti To clarify if you are reffering to the following lines 
https://github.com/unifyai/ivy/blob/d355d7b2c946d87fc8163e72325894d927c3a2b4/ivy/functional/frontends/numpy/indexing_routines/inserting_data_into_arrays.py#L51-L54

This is necessary as they might get updated depending on the first input argument, like here 
https://github.com/unifyai/ivy/blob/d355d7b2c946d87fc8163e72325894d927c3a2b4/ivy/functional/frontends/numpy/indexing_routines/inserting_data_into_arrays.py#L91-L93

File: ivy/functional/frontends/numpy/indexing_routines/inserting_data_into_arrays.py [hunk 4]
```diff
@@ -1,7 +1,9 @@
 import ivy
+
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/numpy/indexing_routines/inserting_data_into_arrays.py:2]:
```suggestion
```

File: ivy/functional/frontends/numpy/matrix/methods.py [hunk 5]
```diff
@@ -23,13 +23,21 @@ def _init_data(self, data, dtype, copy):
                 dtype = data.dtype
             data = ivy.array(data, dtype=dtype, copy=copy)
             self._data = data
+        elif ivy.isscalar(data):
+            self._data = ivy.array(data)
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/numpy/matrix/methods.py:27]: Can we use `ivy.asarray` then cast this scalar value to dtype

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey @ShreyanshBardia 

A few changes to be done then we get the PR merged

File: ivy/functional/frontends/numpy/matrix/methods.py (hunk 5 above)
Comment (ShreyanshBardia) [reply ivy/functional/frontends/numpy/matrix/methods.py]: done

**Review by hirwa-nshuti — APPROVED** [review]
Hey @ShreyanshBardia 

Thanks for this PR😊
All looks good to me and going to get it merged🤞
